x-pack/plugins/observability_solution/observability_ai_assistant/server/service/client/index.ts

```diff
@@ -206,20 +206,21 @@ export class ObservabilityAIAssistantClient {
 
         const next = async (nextMessages: Message[]): Promise<void> => {
           const lastMessage = last(nextMessages);
-
           const isUserMessage = lastMessage?.message.role === MessageRole.User;
 
-          const isUserMessageWithoutFunctionResponse = isUserMessage && !lastMessage?.message.name;
-
-          const contextFirst =
-            isUserMessageWithoutFunctionResponse && functionClient.hasFunction('context');
+          const indexOfLastUserMessage = findLastIndex(
+            nextMessages,
+            ({ message }) => message.role === MessageRole.User && !message.name
+          );
 
-          const isAssistantMessageWithFunctionRequest =
-            lastMessage?.message.role === MessageRole.Assistant &&
-            !!lastMessage?.message.function_call?.name;
+          const shouldInjectContext =
+            indexOfLastUserMessage !== -1 &&
+            nextMessages.slice(indexOfLastUserMessage).every(({ message }) => {
+              return message.function_call?.name !== 'context';
```

Comment:
fwiw, you should be able to move this out of the `next` function, but it's more of a nit/cleanup

Reply:
If I move this out, won't it cause an infinite loop, since `shouldInjectContext` will always be true?

Reply:
No you'd do it outside of `next()`, and you wouldn't call `next()` as part of injecting the function request and response. however I just realized that you need `next()` for creating the function response, so probably better to leave it as-is for now.